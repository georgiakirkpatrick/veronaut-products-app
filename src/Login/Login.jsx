import React, { useState } from 'react'
import { Link, useNavigate } from 'react-router'
import AuthApiService from '../services/auth-api-service'
import FormTitle from '../FormTitle/FormTitle'
import FormButton from '../FormButton/FormButton'
import FormTextInput from '../FormTextInput/FormTextInput'
import TokenService from '../services/token-service'
import bcrypt from "bcryptjs-react"
import './Login.css'

const Login = ({
  routeProps = {
    match: {
      params: {
        categoryId: '1'
      }
    }
  }
}) => {
  const [error, setError] = useState(null)
  const [loginEmail, setLoginEmail] = useState('')
  const [loginPassword, setLoginPassword] = useState('')

  const history = useNavigate()

  const advancePage = () => {
    const referrer = routeProps.location.state
      ? routeProps.location.state.referrer
      : '/'

    history.push(referrer)
  }

  const handleLoginSubmit = event => {
    if (loginEmail === '' && loginPassword === '') {
       alert('Please enter your email and password')
    } else if (loginEmail === '' && loginPassword !== '') {
      alert('Please enter your email')
    } else if (loginEmail !== '' && loginPassword === '') {
      alert('Please enter your password')
    } else {
      event.preventDefault()

      const hashedPassword = pw => {
        const salt = bcrypt.genSaltSync(10)
          const hash = bcrypt.hashSync(pw, salt)

          return hash
      }

      AuthApiService.postLogin({
        email: loginEmail,
        password: hashedPassword(loginPassword)
      })
      .then(response => {
        setLoginEmail('')
        setLoginPassword('')
        TokenService.saveAuthToken(response.authToken)
      })
      .catch(response => {
        setError(response.error)
        console.log('error', error)
      })

      advancePage()
    }
  }

  return (
    <form className='Login'>
      <header>
        <FormTitle titleText='Log in' />
      </header>
      
      <fieldset className='Login__fieldset'>
        <FormTextInput
          id='login-email'
          prompt='Email'
          name='loginEmail'
          currentValue={loginEmail}
          handleChange={event => {setLoginEmail(event.target.value)}}
        />

        <FormTextInput
          id='login-password'
          prompt='Password'
          name='loginPassword'
          currentValue={loginPassword}
          handleChange={event => {setLoginPassword(event.target.value)}}
        />

        <Link to='/forgot-password'>Forgot password?</Link>
      </fieldset>

      <FormButton 
        buttonText='SIGN IN'
        handleClick={event => handleLoginSubmit(event)}
      />

      <Link to={`/create-account`} className='main-link'>
        Create an account
      </Link>
    </form>
  )
}

export default Login