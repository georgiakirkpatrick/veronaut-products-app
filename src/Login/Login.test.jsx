import React from 'react'
import { render, screen } from '@testing-library/react'
import { describe, it, expect } from 'vitest'
import { BrowserRouter as Router } from 'react-router'
import Login from './Login'
import { handleLoginSubmit } from './Login'

describe ('Login', () => {
  it ('renders without crashing', () => {
    render(<Router><Login /></Router>)
    screen.debug()
  })

  it('renders the UI as expected', () => {
    const tree = render(<Router><Login /></Router>)
    expect(tree).toMatchSnapshot()    
  })

    // it('checks that all required fields are supplied', () => {
    //     const loginEmail = 'test@email.com'
    //     const loginPassword = 'testpassword'

    //     handleLoginSubmit()
    // })
})