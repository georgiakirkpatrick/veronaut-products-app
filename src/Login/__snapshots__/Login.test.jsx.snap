// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > renders the UI as expected 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="Login"
      >
        <header>
          <h2
            class="FormTitle"
          >
            Log in
          </h2>
        </header>
        <fieldset
          class="Login__fieldset"
        >
          <div
            class="FormTextInput"
          >
            <label
              for="login-email"
            >
              Email
            </label>
            <input
              id="login-email"
              name="loginEmail"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="FormTextInput"
          >
            <label
              for="login-password"
            >
              Password
            </label>
            <input
              id="login-password"
              name="loginPassword"
              required=""
              type="text"
              value=""
            />
          </div>
          <a
            href="/forgot-password"
          >
            Forgot password?
          </a>
        </fieldset>
        <button
          class="FormButton"
          type="button"
        >
          SIGN IN
        </button>
        <a
          class="main-link"
          href="/create-account"
        >
          Create an account
        </a>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="Login"
    >
      <header>
        <h2
          class="FormTitle"
        >
          Log in
        </h2>
      </header>
      <fieldset
        class="Login__fieldset"
      >
        <div
          class="FormTextInput"
        >
          <label
            for="login-email"
          >
            Email
          </label>
          <input
            id="login-email"
            name="loginEmail"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="FormTextInput"
        >
          <label
            for="login-password"
          >
            Password
          </label>
          <input
            id="login-password"
            name="loginPassword"
            required=""
            type="text"
            value=""
          />
        </div>
        <a
          href="/forgot-password"
        >
          Forgot password?
        </a>
      </fieldset>
      <button
        class="FormButton"
        type="button"
      >
        SIGN IN
      </button>
      <a
        class="main-link"
        href="/create-account"
      >
        Create an account
      </a>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
